jetbrains-core/tst/software/aws/toolkits/jetbrains/services/cwc/controller/ChatControllerTest.kt

```diff
@@ -0,0 +1,794 @@
+            pillText = testPillText,
+            prompt = testPrompt,
+        )
+
+        val testTabId = "testTabId"
+        val testMessageId = "testMessageId"
+        val testCommand = "testCommand"
+        val testMessage = IncomingCwcMessage.FollowupClicked(
+            followUp = testFollowUp,
+            tabId = testTabId,
+            messageId = testMessageId,
+            command = testCommand,
+        )
+
+        val testActiveFileContext = ActiveFileContext(null, null)
+
+        // Mock no auth needed
+        every { authController.getAuthNeededState(any()) } returns null
+
+        // Session History Mock
+        val mockChatSessionInfo = mockk<ChatSessionInfo>(relaxed = true)
+        val testHistory = mutableListOf<ChatRequestData>()
+        every { mockChatSessionInfo.history } returns testHistory
+        every { mockChatSessionInfo.scope } returns CoroutineScope(Job())
+        every { chatSessionStorage.getSession(any(), any()) } returns mockChatSessionInfo
+
+        val testUserIntent = UserIntent.EXPLAIN_CODE_SELECTION
+        every { intentRecognizer.getUserIntentFromFollowupType(any()) } returns testUserIntent
+
+        // Act
+        runBlocking {
+            chatController.processFollowUpClick(testMessage, testTriggerId)
+        }
+
+        // Assert
+        val testRequestData = ChatRequestData(
+            tabId = testTabId,
+            message = testPrompt,
+            activeFileContext = testActiveFileContext,
+            userIntent = testUserIntent,
+            triggerType = TriggerType.Click,
+        )
+
+        assertThat(
+            testHistory[0] == testRequestData
+        ).isTrue()
+
+        verify { telemetryHelper.recordEnterFocusConversation(match { it == testTabId }) }
+        verify { telemetryHelper.recordStartConversation(match { it == testTabId }, match { it == testRequestData }) }
+
+        verify {
+            handlerMock.handle(
+                match { it == testTabId },
+                match { it == testTriggerId },
+                match { it == testRequestData },
+                match { it == mockChatSessionInfo }
+            )
+        }
+
+        coVerify { telemetryHelper.recordInteractWithMessage(match { it == testMessage }) }
+    }
+
+    @Test
+    fun `processCodeWasCopiedToClipboard handles code copy`() {
+        // Arrange
+        val testMessage = IncomingCwcMessage.CopyCodeToClipboard(
+            tabId = "testTabId",
+            messageId = "testMessageId",
+            command = "testCommand",
+            code = "testCode",
+            insertionTargetType = "testInsertionTargetType",
+        )
+
+        // Act
+        runBlocking {
+            chatController.processCodeWasCopiedToClipboard(testMessage)
+        }
+
+        // Assert
+        coVerify { telemetryHelper.recordInteractWithMessage(match { it == testMessage }) }
+    }
+
+    @Test
+    fun `processInsertCodeAtCursorPosition handles code insertion`() {
```

Comment:
This tests is excessively long, with a lot of mocking. That is a pretty good indicator that the underlying code should be refactored.

Reply:
How would you propose refactoring the underlying code though?  The actual `processInsertCodeAtCursorPosition` method is about 30 lines of code, but its reliance on various editor/static classes which thereby increase the complexity.  

The scope of the underlying method itself also doesn't seem too big - it's main functionality is to insert code at cursor position and not much else. 

If I were to break that underlying method into several methods, those methods should be private (so unsure how to mock).  If it is possible to mock them and then to write separate tests for them that could reduce the scope here, but unsure if that's even possible.